### ISSUE TITLE
[data] [streaming] Limit operator shouldn't materialize stream

### ISSUE THREAD

**Issue by ericl** [ray-project/ray#34234]
The following shouldn't materialize all previous blocks:

```
ray.data.range(1000, parallelism=1000).map(lambda x: x).limit(1).materialize()
```

This can be implemented with a special Limit operator that terminates once it receives enough elements.

**Comment by c21** [comment]
This will work with https://github.com/ray-project/ray/issues/32761 .

**Comment by raulchen** [comment]
@c21 @ericl 

#34705 implements the Limit physical operator. But it hasn't implemented the logic to prevent range from producing more data than needed by limit. There are 2 ways to do this:
1. Make the LimitOperator return a flag to indicate that the limit is reached. The executor should check this flag and cancel upstream operators.
2. Push down the limit to the source read operator, when optimizing the logical plan.

2 is the common practice for most data processing systems, and better for long term. But I'm not sure how close we are to achieve this. If not, we may want to implement 1 first.

**Comment by ericl** [comment]
It sounds like (2) may not be available until after 2.5. I think we need (1) anyways for the general case (e.g., limit after filter), so how about starting with that one?
